### ISSUE TITLE
flock error on lustre file system

### ISSUE THREAD

**Issue by sillitoe** [dagolden/Path-Tiny#211]
I've encountered the following error running `path('file.txt')->slurp` in an HPC environment. 

```
Error flock (1) on 'file.txt': Function not implemented
```

Which comes from:

https://github.com/dagolden/Path-Tiny/blob/80c3eca4bc3eca61e3429047d55597e32446e7f8/lib/Path/Tiny.pm#L1078

Seems to stem from trying to slurp a file on a lustre file system.

All existing tests work fine:

```bash
$ pwd
/lustre/scratch/scratch/ucbcisi
$ git clone https://github.com/dagolden/Path-Tiny.git
$ cd Path-Tiny
$ prove -l -r t | tail -n 1
Result: PASS
```

But if I create and run the following...

```perl
# t/slurp_lustre.t
use Test::More tests => 2;

use strict;
use warnings;
use FindBin;

use Path::Tiny;
use File::Copy qw/ copy /;

my $lustre_path = path("$FindBin::Bin/data/chmod.txt");
my $non_lustre_path = path( "/tmp/tmp.txt" );
copy( $lustre_path, $non_lustre_path ) or die "error: $!";

is( length($non_lustre_path->slurp), 50353, "length looks okay for '$non_lustre_path'" );
is( length($lustre_path->slurp),     50353, "length looks okay for '$lustre_path'" );
```

Then the non-lustre path is fine; the lustre path gives the expected error:

```bash
$ prove -l -v t/slurp_lustre.t
t/slurp_lustre.t .. 
1..2
ok 1 - length looks okay for '/tmp/tmp.txt'
Error flock (1) on '/lustre/scratch/scratch/ucbcisi/Path-Tiny/t/data/chmod.txt': Function not implemented at t/slurp_lustre.t line 17.
```

I doubt I'm going to be able to get an upstream fix that makes flock available - is there any way of explicitly asking `Path::Tiny` **not** to use flock even if the system says it is available?

At the moment, I'm using the following hack:

```perl
diff --git a/lib/Path/Tiny.pm b/lib/Path/Tiny.pm
index e179b21..b8fe9f8 100644
--- a/lib/Path/Tiny.pm
+++ b/lib/Path/Tiny.pm
@@ -50,6 +50,7 @@ sub _check_PU {
     !!eval { require PerlIO::utf8_strict; PerlIO::utf8_strict->VERSION(0.003); 1 };
 }
 
+our $NO_FLOCK = 0;
 my $HAS_FLOCK = $Config{d_flock} || $Config{d_fcntl_can_lock} || $Config{d_lockf};
 
 # notions of "root" directories differ on Win32: \\server\dir\ or C:\ or \
@@ -1025,7 +1026,7 @@ sub filehandle {
     $binmode = "" unless defined $binmode;
 
     my ( $fh, $lock, $trunc );
-    if ( $HAS_FLOCK && $args->{locked} ) {
+    if ( !$NO_FLOCK && $HAS_FLOCK && $args->{locked} ) {
         require Fcntl;
         # truncating file modes shouldn't truncate until lock acquired
         if ( grep { $opentype eq $_ } qw( > +> ) ) {
```

And then setting

```
$Path::Tiny::NO_FLOCK=1
```

Happy to turn that into a PR if that's useful?

Ta.

System info:

```
$ grep /lustre /etc/fstab 
<...>:/lustre01 /lustre lustre defaults,_netdev,user_xattr 0 0

$ perl -V
Summary of my perl5 (revision 5 version 22 subversion 0) configuration:
   
  Platform:
    osname=linux, osvers=3.10.0-123.20.1.el7.x86_64, archname=x86_64-linux-thread-multi
    uname='linux login99.dev.legion.ucl.ac.uk 3.10.0-123.20.1.el7.x86_64 #1 smp wed jan 21 09:45:55 est 2015 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0 -Duseshrplib -Dusethreads -Aeval:scriptdir=/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='4.8.2 20140120 (Red Hat 4.8.2-16)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /shared/ucl/apps/git/2.3.5/gnu-4.9.2/lib64/../lib64 /lib/../lib64 /usr/lib/../lib64 /shared/ucl/apps/subversion/1.8.13/lib /shared/ucl/apps/apr-util/1.5.4/lib /shared/ucl/apps/apr/1.5.2/lib /shared/ucl/apps/git/2.3.5/gnu-4.9.2/lib64 /lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.17.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.17'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/5.22.0/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options: HAS_TIMES MULTIPLICITY PERLIO_LAYERS
                        PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD
                        PERL_IMPLICIT_CONTEXT PERL_MALLOC_WRAP
                        PERL_NEW_COPY_ON_WRITE PERL_PRESERVE_IVUV
                        USE_64_BIT_ALL USE_64_BIT_INT USE_ITHREADS
                        USE_LARGE_FILES USE_LOCALE USE_LOCALE_COLLATE
                        USE_LOCALE_CTYPE USE_LOCALE_NUMERIC USE_LOCALE_TIME
                        USE_PERLIO USE_PERL_ATOF USE_REENTRANT_API
  Locally applied patches:
        Devel::PatchPerl 1.34
  Built under linux
  Compiled at Jul  7 2015 17:08:21
  %ENV:
    PERL5LIB="/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0:/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl:/shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib"
  @INC:
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0/x86_64-linux-thread-multi
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0/x86_64-linux-thread-multi
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/5.22.0/x86_64-linux-thread-multi
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/5.22.0
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0/x86_64-linux-thread-multi
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/site_perl/5.22.0
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/5.22.0/x86_64-linux-thread-multi
    /shared/ucl/apps/perl/perlbrewroot/perls/perl-5.22.0/lib/5.22.0
    .
```

**Comment by xdg** [comment]
The problem of knowing what a filesystem supports is tricky.  My inclination is the following:
* Allow an environment variable "PERL_PATH_TINY_NO_FLOCK" to override the normal `$HAS_FLOCK` detection.  That's easier to apply universally than setting a global variable.
* Generalize the BSD-specific logic in `_throw` so that any flock attempt that matches "operation not supported" -- or, in your case, "Function not implemented" -- warns once but is not normally fatal.

I think that's a reasonable compromise -- by default, failing flock gets a single warning; one can optionally fatalize it; or one can disable flocking entirely with an environment variable.

I'd be happy to look at a PR that takes that approach!  Or if that's too much for you, I'm willing to write the code, but would need your commitment to test it with the lustre filesystem.

**Comment by sillitoe** [comment]
Thanks for getting back so quickly. No problem, sorting a PR now.

I guess one possible issue with setting `$HAS_FLOCK` from an environment variable is that it wouldn't be possible to make changes after startup. I can't thinking of a reason why someone *would* want to change this after startup, but I guess they might. Anyway - easy to adjust later on if necessary - the exception handling will solve my case anyway.

I'm happy to test this against lustre, however I don't think I'll be able to set up an automated build.

**Comment by xdg** [comment]
I suppose a "no flock" environment variable could be *checked* where `$HAS_FLOCK` is checked in `sub filehandle` -- then it can be switched on/off as needed.  I like that!

No automated build is necessary -- just checking that the behavior is as you expect.

**Comment by sillitoe** [comment]
Sorry, just realised that you might be waiting for me to confirm this works before releasing (it does - thanks very much!)

**Comment by xdg** [comment]
This was fixed in #212.  Closing.
